proposals/lip-0049.md

```diff
@@ -223,135 +223,171 @@ When processing a cross-chain message `ccm`, the functions below may be called b
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = { "ccmID": ccmID, "status": eventStatus},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
         )
 
-    elif crossChainCommandID is not supported:
-        sendInternal(
+        send(
             EMPTY_FEE_ADDRESS,
             ccm.moduleID,
             ccm.crossChainCommandID,
             ccm.sendingChainID,
             0,
-            CCM_STATUS_CROSS_CHAIN_COMMAND_NOT_SUPPORTED,
+            errorStatus,
             ccm.params
         )
+    # Else, we discard the message.
+    else:
+        emitEvent(
+            moduleID = MODULE_ID_INTEROPERABILITY,
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_COMMAND_DISCARDED},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
 ```
 
 #### forward
 
-The `forward` function forwards a CCM to the specified receiving chain. This function is only called on the mainchain when processing a CCU.
+The `forward` function forwards a CCM to the specified receiving chain. This function is only called on the mainchain when processing a CCU. If an error occurs, the CCM is bounced back to the sending chain or discarded.
 
 ```python
-# The ccu including this message is passed as context to this function
-forward(ccu, ccm):
-    if (chainAccount(ccm.receivingChainID) exists
-        and chainAccount(ccm.receivingChainID).status == CHAIN_ACTIVE
-        and isLive(ccm.receivingChainID):
-        # The Token module is called to transfer the fee from sending to receiving chain
-        token.forwardMessageFee(ccm)
-        if the previous call fails or returns False:
-            return
+def forward(ccu: CCU, ccm: CCM) -> None:
+    # Calculate CCM ID, used later in events.
+    ccmID = SHA-256(encode(crossChainMessageSchema, ccm))
+    
+    # Chain with status REGISTERED do not pass this check.
+    if isLive(ccm.receivingChainID) and chainAccount(ccm.receivingChainID).status == CHAIN_STATUS_ACTIVE:
+        # Transfer the fee from escrow to escrow, call Token LIP
+        Token.transferBetweenEscrows(ccm.sendingChainID, ccm.receivingChainID, feeTokenID, fee)
+
         addToOutbox(ccm.receivingChainID, ccm)
+        # Emit CCM forwarded event.
+        emitEvent(
+            moduleID = MODULE_ID_INTEROPERABILITY,
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_FORWARDED},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
         return
 
+    # If ccm status is not ok, stop the processing of this message
     if ccm.status != CCM_STATUS_OK:
-        # Stop the processing of this message
+        # Emit CCM discarded event.
+        emitEvent(
+            moduleID = MODULE_ID_INTEROPERABILITY,
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_DISCARDED},
```

Comment:
Maybe here we could be more fine-grained regarding the error-code so it is clear why the message was discarded (i.e., forwarding failed as receiving chain not active in this case).

Reply:
done